text/0000-promise.md

```diff
@@ -0,0 +1,260 @@
+
+```js
+const util = require('util');
+const native = require('../native');
+
+export const fibonacciAsync = util.promisify(native.fibonacci_async);
+```
+
+Alternatively, a `Promise` can be constructed directly in Neon. Unlike an asynchronous method that accepts a callback, a method that returns a promise requires two parts:
+
+* A `Promise` to return from the method
+* A hook to resolve or reject the `Promise`
+
+In JavaScript, this looks like the following:
+
+```js
+function asyncMethod() {
+    return new Promise((resolve, reject) => {
+        if (Math.random() > 0.5) {
+            resolve();
+        } else {
+            reject();
+        }
+    });
+}
+```
+
+In `Neon`, when a `JsPromise` is constructed, a `Deferred` object is also provided. It can be thought of as the following pattern in JavaScript:
+
+```js
+function deferred() {
+    const deferred = {};
+    const promise = new Promise((resolve, reject) => {
+        deferred.resolve = resolve;
+        deferred.reject = reject;
+    });
+
+    return [promise, deferred];
+}
+
+function asyncMethod() {
+    const [promise, d] = deferred();
+
+    setTimeout(() => d.resolve(), 5000);
+
+    return promise;
+}
+```
+
+This could be written in Neon with the following:
+
+```rust
+fn async_method(cx: FunctionContext) -> JsResult<JsPromise> {
+    let queue = cx.queue();
+    let (promise, d) = cx.promise();
+
+    std::thread::spawn(move || {
+        std::thread::sleep(std::time::Duration::from_millis(5000));
+
+        queue.send(|cx| d.resolve(cx.undefined()));
+    });
+
+    Ok(promise)
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `JsPromise` API consists of two structs:
+
+* `JsPromise`. Opaque value; only useful for passing back to JavaScript.
+* `Deferred`. Handle for resolving and rejecting the related `Promise`.
+
+They may only be created with the `deferred` method on `Context`.
+
+```rust
+trait Context {
+    fn deferred(&mut self) -> (Handle<JsPromise>, Deferred);
+}
+```
+
+## `JsPromise`
+
+Opaque handle that represents a JavaScript `Promise`. It is not threadsafe.
+
+```rust
+/// A JavaScript `Promise`
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct JsPromise(raw::Local);
+
+impl Value for JsPromise {}
+
+impl Managed for JsPromise {
+    fn to_raw(self) -> raw::Local { self.0 }
+    fn from_raw(h: raw::Local) -> Self { JsPromise(h) }
+}
+
+impl ValueInternal for JsPromise {
+    fn name() -> String { "promise".to_string() }
+    fn is_typeof<Other: Value>(env: Env, other: Other) -> bool {
+        unsafe { neon_runtime::tag::is_promise(env.to_raw(), other.to_raw()) }
+    }
+}
+
+unsafe impl This for JsPromise {
+    fn as_this(_env: Env, h: raw::Local) -> Self {
+        JsPromise(h)
+    }
+}
+
+impl JsPromise {
+    pub(crate) fn new_internal<'a>(value: raw::Local) -> Handle<'a, JsPromise> {
+        Handle::new_internal(JsPromise(value))
+    }
+}
+```
+
+## `Deferred`
```

Comment:
I know the JS world has a historical legacy of "Deferred" being the name for this data structure, but I don't know if that history is being lost over time since the ecosystem now uses standard ES promises, which don't have any concept by this name.

And without the mindshare, I'm not sure the name really conveys much intuitive meaning. Is it worth thinking about alternatives? Maybe something that leans on the "sender" precedent with Rust channels? "Fulfiller"? "Completer"? Those are awkward. I don't have any amazing suggestions right now, but I'll think about it.

Reply:
@dherman What do you think about calling it `Sender` to match channels? It's conceptually very similar to a `oneshot`: https://docs.rs/futures/0.3.5/futures/channel/oneshot/index.html.

With that change, instead of separate `resolve` / `reject`, I would opt for a single combined method:

```rust
impl Sender {
    pub fn send<T: Value, E: Value>(result: Result<T, E>) {
        /* ... */
    }
}
```

The downside of this approach is that it requires the user to provide an `Err` type even when the code is infallible, which isn't very ergonomic.